Closes #1319: Conceptualize and featurize toolbar autocompletion

grigoryk opened the pull request:
First pass at this. Take a look at the sample browser change - in the end, it's trivial to wire in toolbar autocompletion with browsing history as a autocompletion provider as the primary source, and define backup data sources.

grigoryk commented:
The "WIP" bit is that I need to update any tests this broke, and likely write new ones. Other than that, it works and feels pretty well.

csadilek commented on components/browser/domains/src/main/java/mozilla/components/browser/domains/Providers.kt:
```diff
@@ -0,0 +1,114 @@
+/**
+ * Provides common autocomplete functionality powered by domain lists.
+ */
+abstract class Provider : Autocomplete {
```
I'd find `AutocompleteProvider` and `DomainAutocompleteProvider` (for the common subtype) easier to read.

csadilek commented on components/browser/domains/src/main/java/mozilla/components/browser/domains/Providers.kt:
```diff
@@ -0,0 +1,114 @@
+
+    abstract fun initialize(context: Context)
+
+    // We compute these on worker thread; make sure results are immediately visible on the UI thread.
```
Looks like a left-over comment.

csadilek commented on components/concept/toolbar/src/main/java/mozilla/components/concept/toolbar/Autocomplete.kt:
```diff
@@ -0,0 +1,40 @@
+
+    /**
+     * Describes an autocompletion result.
+     * @param text Result of autocompletion, text to be displayed.
```
nit: these are `@property`

grigoryk commented:
@csadilek take another look when you get a chance: updated tests, renamed/cleaned up stuff, kept 'source' and 'url' in the result to make focus update straightforward.

grigoryk commented:
TC tasks failed because of codecov service outage: https://github.com/mozilla-mobile/android-components/issues/1364

pocmo commented on components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarAutocompleteFeature.kt:
```diff
@@ -0,0 +1,28 @@
+class ToolbarAutocompleteFeature(val toolbar: Toolbar, private val providers: List<AutocompleteProvider>) {
+    init {
+        toolbar.setAutocompleteListener { value, acceptor ->
+            acceptor?.let {
```
I wonder if we should ever end up here with a `null` acceptor. This seems to be unneeded work. I remember that this may come from InlineAutocomplete; I wonder if we should stop that event ("autocomplete for nobody") as early as possible. Doesn't need to be here though.

grigoryk replied:
Yeah, that crept in from some existing code, and I didn't clean it up.

grigoryk commented:
We've discussed this offline, and option (B) as proposed by @pocmo is worth the effort to re-align this patch series.

grigoryk commented:
@pocmo take another look, this is following your suggested pattern (B). A bit of boilerplate here and there, and I've kept part of the `concept-toolbar` changes (the "acceptor" side of things) to make interacting with the ui components simpler.

csadilek commented on components/concept/toolbar/src/main/java/mozilla/components/concept/toolbar/Toolbar.kt:
```diff
@@ -58,6 +58,14 @@ interface Toolbar {
+     * @param filter A functions which is will perform autocompletion and send results to
+     * [AutocompleteProvider.ResultAcceptor].
+     */
+    fun setAutocompleteListener(filter: (String, AutocompleteAcceptor?) -> Unit)
```
We need to update the parameter name now.

csadilek commented on components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarAutocompleteFeature.kt:
```diff
@@ -0,0 +1,69 @@
+    // an "AutocompleteProvider" concept. This pattern allows a storage component, for example, to
+    // be standalone and not depend on something like 'concept-domains'.
+    // This pattern also helps prevent circular dependency situations from arising over time.
+
```
I get the reason we're doing this, but it is a bit unfortunate :). Wondering if we should create a single AutocompleteResult (instead of 3-4 distinct types) in a root module e.g. support-base? I think this would simplify this quite a bit!

grigoryk replied:
@csadilek One of the reasons *for* doing this boilerplate that's not currently obvious is that it enables individual providers to define a "result" as they please, and expose a greater set of functionality as they see fit. The `feature` component is then responsible for massaging different results together. If we have a common definition in `support-*`, we introduce an implicit dependency between all of our providers.

The other side of having this implicit dependency is that it has a benefit of imposing a "growing in lockstep" limitation on the different provider implementations, which could be quite useful.

pocmo replied:
Boilerplate is the price for abstractions and a layered architecture. This seems fine to me here since Features are translating between two worlds. Anyhow, boilerplate is something we can always optimize later. :)

csadilek replied:
Yeah, if the results were different I wouldn't have noticed. It's just that it's the same type with 4 copies, which tipped me off. So, this boilerplate is caused by a lack of abstraction (over the result) not because of it. :)

Anyways, like I wrote, I don't feel strongly. Just a comment...

pocmo commented on components/ui/autocomplete/build.gradle:
```diff
@@ -27,6 +27,7 @@ dependencies {
     implementation Dependencies.kotlin_stdlib
 
     implementation Dependencies.support_appcompat
+    implementation project(':concept-toolbar')
```
I think we should avoid having this dependency too:

* `ui-*` components are dumb view components without any knowledge about the context they are used in. This ties this to the toolbar.

* The autocomplete component can be used for autocompleting all kinds of things. You could throw it into some app and autocomplete phone numbers or user names. However the concept leaks the idea of "URLs" into it now.

`BrowserToolbar` accepts a an `AutocompleteAcceptor`and in `setAutocompleteListener()` it could create a wrapper/delegate that implements `AutocompleteAcceptor` and delegates to `InlineAutocompleteEditText`. What do you think?

grigoryk replied:
I think that's OK. This creates an annoying delegate chain, and I really, really, really dislike delegate chains (you can blame Fennec sync for that...). But it doesn't feel very out of place here, as long as we make sure this doesn't set a wider precedent for these damn things ;)

grigoryk commented:
Failed an unrelated test due to an OOM, see https://github.com/mozilla-mobile/android-components/issues/1467.

pocmo commented:
I tried to fix the conflict and land this, but I'm seeing some other unresolved reference issues after rebasing. :(
